### PR TITLE
feat: tab completion for citation keys

### PR DESCRIPTION
**Pull request by wjiayis**
#33 

In short, it'll

1. [Frontend] Recognize that user is trying to add a citation (trigger text is `\cite{`)
2. [Frontend] Temporarily suppress default Overleaf dropdown suggestions
3. [Frontend] Get the last sentence as context for LLM
4. [Backend] Fetch bibliography in `.bib` files as raw text, and remove irrelevant fields to save tokens
5. [Backend] Call XtraMCP to get paper abstract, using paper title as key
6. [Backend] Query a fast LLM (hardcoded to `gpt-5.2` for now) to get at most 3 citation keys
7. [Frontend] Suppress default Overleaf tab-completion to allow users to accept citation suggestions

### CODE REVIEW

**Comment by Junyi-99** [comment]
Hi @wjiayis, thanks for the update. I’ve created a new issue to address the token expiration problem.

Regarding the latency issue, do we have visibility into which part of the pipeline contributes most to the high latency? For example, a rough breakdown across:

`frontend → backend → LLM provider (reasoning + response)`

I’ll take a look at this PR later this evening as well.

**Comment by wjiayis** [comment]
With regards to latency, just changing the model fixed the bulk of the issue. I still implemented prompt caching and prompt refinement, but did not implement the "no-reply" query as performance improvement would not be much and it'd overcomplicate the feature.

File: internal/services/toolkit/client/get_citation_keys.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
+package client
+
+// TODO: This file should not place in the client package.
```
[the remark this answers is not in the document]
Comment (wjiayis) [reply internal/services/toolkit/client/get_citation_keys.go:3]: @Junyi-99 I'm planning to eventually move this file to wherever `internal/services/toolkit/client/get_conversation_title_v2.go` will be moved to, since both files contain high-order business logic that doesn't belong to the `client` folder. Shall I leave this file here for now?

[the remark this answers is not in the document]
Comment (4ndrelim) [reply internal/services/toolkit/client/get_citation_keys.go:3]: i got copilot to review the PR. I didnt really look through its suggestions. Could you resolve copilot's comments and integrate any valid / viable suggestions? I think jun yi is busy these few days, i'll review later on this week if its still pending review! Thanks for your hardwork jiayi! 

@Junyi-99 actually do we have a recommended PR to main workflow? Should we review only review PRs to main and allow freely merging to staging / dev branch for testing in a simulated prod env? Or should there be a preliminary review before merging to staging too?

File: proto/chat/v2/chat.proto [hunk 2]
```diff
@@ -234,3 +240,16 @@ message CreateConversationMessageStreamResponse {
...
+message GetCitationKeysRequest {
+  string sentence = 1;
+  string project_id = 2;
+  optional string model_slug = 3;
+}
+
+// Response containing the suggested citation keys
+message GetCitationKeysResponse {
+  // A comma-separated string of keys, or empty if none found
+  string citation_keys = 1;
```
[the remark this answers is not in the document]
Comment (Junyi-99) [reply proto/chat/v2/chat.proto]: @wjiayis I agree with Copilot’s suggestion: citation_keys should be a repeated string rather than a comma-separated string

File: webapp/_webapp/src/views/settings/sections/beta-feature-settings.tsx [hunk 3]
```diff
@@ -9,8 +9,8 @@ export const BetaFeatureSettings = () => {
     <SettingsSectionContainer>
       <SettingsSectionTitle>Beta Features</SettingsSectionTitle>
       <SettingItem
-        label="Enable completion"
-        description="Predicts your next edit"
+        label="Enable citation suggestions"
+        description="Suggest citations as you write"
         isLoading={isUpdating.enableCompletion}
```
Comment (Junyi-99) [inline webapp/_webapp/src/views/settings/sections/beta-feature-settings.tsx]: The property name `enableCompletion` should be renamed to `enableCitationSuggestions` to match the updated label `Enable citation suggestions`. 

Additionally, you'll likely need to update the corresponding field name in the `message Settings` definition in `proto/user/v1/user.proto` to maintain consistency across the codebase.

File: internal/services/toolkit/client/get_citation_keys.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	// Normalize multiple spaces
+	multiSpaceRe := regexp.MustCompile(` {2,}`)
+	bibliography = multiSpaceRe.ReplaceAllString(bibliography, " ")
+
+	return bibliography, nil
+}
+
+func (a *AIClientV2) GetCitationKeys(ctx context.Context, sentence string, userId bson.ObjectID, projectId string, llmProvider *models.LLMProviderConfig) (string, error) {
+	bibliography, err := a.GetBibliographyForCitation(ctx, userId, projectId)
+
```
[the remark this answers is not in the document]
Comment (wjiayis) [reply internal/services/toolkit/client/get_citation_keys.go]: Added `internal/services/toolkit/client/get_citation_keys_test.go`. @4ndrelim @Junyi-99 What's the typical workflow for writing test cases?

File: internal/services/toolkit/client/get_citation_keys.go [hunk 5]
```diff
@@ -5,106 +5,177 @@ import (
...
+}
 
-			bibLines = append(bibLines, line)
+// fetchAbstracts enriches entries with abstracts from XtraMCP in parallel.
+func (a *AIClientV2) fetchAbstracts(ctx context.Context, entries []string) []string {
+	result := make([]string, len(entries))
+	copy(result, entries)
+
+	loader := xtramcp.NewXtraMCPLoaderV2(nil, nil, a.cfg.XtraMCPURI)
+	sem := make(chan struct{}, MAX_CONCURRENT_XTRAMCP)
```
Comment (wjiayis) [inline internal/services/toolkit/client/get_citation_keys.go]: Points for future optimizations:
1. Create a batch API in XtraMCP to reduce number of calls.
2. Cache outputs to avoid calling XtraMCP every citation suggestion (will leave as TODO for next time).

File: proto/chat/v2/chat.proto [hunk 6]
```diff
@@ -31,6 +31,12 @@ service ChatService {
   rpc ListSupportedModels(ListSupportedModelsRequest) returns (ListSupportedModelsResponse) {
     option (google.api.http) = {get: "/_pd/api/v2/chats/models"};
   }
+  rpc GetCitationKeys(GetCitationKeysRequest) returns (GetCitationKeysResponse) {
+    option (google.api.http) = {
+      post: "/_pd/api/v2/chats/citation-keys"
```
Comment (Junyi-99) [inline proto/chat/v2/chat.proto]: The `GetXXX` request should use `get: ...` instead of `post: ...`

File: webapp/_webapp/src/query/api.ts [hunk 7]
```diff
@@ -145,6 +147,13 @@ export const getProject = async (data: PlainMessage<GetProjectRequest>) => {
   return fromJson(GetProjectResponseSchema, response);
 };
 
+export const getCitationKeys = async (data: PlainMessage<GetCitationKeysRequest>) => {
+  const response = await apiclientV2.post(`/chats/citation-keys`, data, {
```
Comment (Junyi-99) [inline webapp/_webapp/src/query/api.ts]: use `get` instead of `post`. as described in `chat.proto` line `36`

File: internal/services/toolkit/client/get_citation_keys.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Junyi-99) [reply internal/services/toolkit/client/get_citation_keys.go:3]: @4ndrelim thanks for the reviewing, I've created a new issue to detail th workflows. https://github.com/PaperDebugger/paperdebugger/issues/116

[the remark this answers is not in the document]
Comment (Junyi-99) [reply internal/services/toolkit/client/get_citation_keys.go:3]: @wjiayis Hi Jiayi, we’ve left the file here for now. I wasn’t sure how to move it.

**Review by Junyi-99 — CHANGES_REQUESTED** [review]
@wjiayis Hi Jiayi, thanks for your patience. I’ve left some comments, once those are addressed, it can be merged.

File: internal/services/toolkit/client/get_citation_keys.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (4ndrelim) [reply internal/services/toolkit/client/get_citation_keys.go]: We don't have a fixed structure. Currently, we stick to the standard [convention](https://go.dev/doc/tutorial/add-a-test) in Go to place `_test.go` files in the same directory as the source code files they are testing.

Yeah, i think its wise to test involved logic like citation or bib parsing like you did. Unit test cases to test isolated logic and edge cases and maybe 1-2 integrated test cases chaining the functions to test higher order logic.

File: internal/services/toolkit/client/get_citation_keys.go (hunk 5 above)
Comment (4ndrelim) [reply internal/services/toolkit/client/get_citation_keys.go]: nice. yeah, i think batch API would be more efficient. We can create a separate batch endpoint on XtraMCP's end.

Edit: I saw you defined `MAX_CONCURRENT_XTRAMCP` as 24. Just curious how u decided the number. But i think it should be fine since its only read and psql handles read concurrency well.

File: internal/services/toolkit/tools/xtramcp/loader_v2.go [hunk 8]
```diff
@@ -237,3 +248,31 @@ func (loader *XtraMCPLoaderV2) fetchAvailableTools() ([]ToolSchemaV2, error) {
 
 	return mcpResponse.Result.Tools, nil
 }
+
+// GetPaperAbstract fetches the abstract for a paper given its title via the REST API
+func (loader *XtraMCPLoaderV2) GetPaperAbstract(ctx context.Context, title string) (*PaperAbstractResponse, error) {
```
Comment (4ndrelim) [inline internal/services/toolkit/tools/xtramcp/loader_v2.go]: I'm thinking if it might be better to put this in a separate file. We can still park this logic in `tools/xtramcp` but `loader` is more concerned with loading and intialization of xtramcp backend services. `GetPaperAbstract` is a service-level logic so maybe a different services file might be more apt.

Also, just a quick check, your logic doesn't require `sessionID` right? No need for MCP tool calling right, just the DB access managed by XtraMCP backend? Yup, then creating a new loader (but not initializing) is correct. If in the future you need the `sessionID`, should use the cached / created upon initialization.

Comment (wjiayis) [reply internal/services/toolkit/tools/xtramcp/loader_v2.go]: Yup this endpoint is not for MCP tool call, just a REST API. I refactored this part and isolated it into `tools/xtramcp/services.go`.

File: internal/services/toolkit/client/get_citation_keys.go (hunk 5 above)
Comment (wjiayis) [reply internal/services/toolkit/client/get_citation_keys.go]: Creating batch endpoint now; 24 is random haha, just to speed up the querying while not having too much concurrency

**Comment by wjiayis** [comment]
All comments are resolved! I'll work with @4ndrelim tomorrow to test and merge the PR.
cc: @Junyi-99
